### Advent of Code: Day 1

#### My process: 
We're summing the digits found at [this link](https://adventofcode.com/2018/day/1/input). I tried to directly pull in the link, but it didn't go very well. Then, I downloaded the input as a `.txt` file and saved it in the repo and opened that directly!

In [1]:
# Open the file for reading.
file = open('./day_1_input.txt', 'r')

# List comprehension iterating over each item in the file
print(sum([int(i) for i in file]))

513


#### How would I otherwise do this?
In general, I probably would have defaulted to importing Pandas and reading the file in using a `.read_table` method or something like that.

#### How might I improve this?
If the list of numbers was **really** large, I would consider parallelizing it. Since we're just adding lots of numbers, I could take my big list of numbers, split it into two lists, have one computer core add the first list and have a second computer core add the second list at the same time. Then, I can just add the two results together to get my final result. I'm not limited to two cores; I could split my list up into lots of different sub-lists. It takes some time to set up, but since my cores can do the adding simultaneously, if I'm adding trillions of numbers, I can cut down how much processing time this will take.

It won't be worth it in this situation, but let me check... how large is my list?

In [2]:
print(len([int(i) for i in file]))

0


This is weird. Why am I getting a 0?

In [3]:
for i in file:
    print(i)

Still weird. I'm not getting anything. Let me literally copy what I wrote above.

In [4]:
# Open the file for reading.
file = open('./day_1_input.txt', 'r')

# List comprehension iterating over each item in the file
print(sum([int(i) for i in file]))

513


This works!

In [5]:
# List comprehension iterating over each item in the file
print(sum([int(i) for i in file]))

0


This doesn't. What this means is that I need to define `file` in my same cell - otherwise it won't work!

In [6]:
# Open the file for reading.
file = open('./day_1_input.txt', 'r')

# List comprehension iterating over each item in the file
print(len([i for i in file]))

1016


My whole list was 1,016 items long. It certainly isn't worth parallelizing... but that's a strategy I'd use if I'm adding way more items.

### Resources Used
- https://stackoverflow.com/questions/32971752/python-read-file-from-a-web-url
- https://www.pythonforbeginners.com/cheatsheet/python-file-handling